internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,229 @@
+}
+
+// collectResources collects Resource objects from the data source, persists them to the database,
+// and signals any change events to the notification processor.
+func (c *Collector) collectResources(ctx context.Context, dataSource DataSource,
+	pools []models.ResourcePool) ([]models.Resource, error) {
+	resources, err := dataSource.GetResources(ctx, pools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resources: %w", err)
+	}
+
+	// Loop over the set of resources and create the associated resource types.
+	seen := make(map[uuid.UUID]bool)
+	for _, resource := range resources {
+		resourceType, err := dataSource.MakeResourceType(&resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make resource type from '%v': %w", resource, err)
+		}
+
+		if _, ok := seen[resourceType.ResourceTypeID]; ok {
+			// We have already seen this one so skip
+			continue
+		}
+		seen[resourceType.ResourceTypeID] = true
+
+		dataChangeEvent, err := persistObjectWithChangeEvent(
+			ctx, c.repository.Db, *resourceType, resourceType.ResourceTypeID, nil, func(object interface{}) any {
+				record, _ := object.(models.ResourceType)
+				return models.ResourceTypeToModel(&record)
+			})
+		if err != nil {
+			return nil, fmt.Errorf("failed to persist resource type': %w", err)
+		}
+
+		if dataChangeEvent != nil { // nolint: staticcheck
+			// TODO: notify the notification processor of the new data change event via a channel
+		}
+	}
+
+	// Loop over the set of resources and insert (or update) as needed
+	for _, resource := range resources {
+		dataChangeEvent, err := persistObjectWithChangeEvent(
+			ctx, c.repository.Db, resource, resource.ResourceID, &resource.ResourcePoolID, func(object interface{}) any {
+				record, _ := object.(models.Resource)
+				return models.ResourceToModel(&record, nil)
+			})
+		if err != nil {
+			return nil, fmt.Errorf("failed to persist resource: %w", err)
+		}
+
+		if dataChangeEvent != nil { // nolint: staticcheck
+			// TODO: notify the notification processor of the new data change event via a channel
+		}
+	}
+
+	return resources, nil
+}
+
+// collectResourcePools collects ResourcePool objects from the data source, persists them to the database,
+// and signals any change events to the notification processor.
+func (c *Collector) collectResourcePools(ctx context.Context, dataSource DataSource) ([]models.ResourcePool, error) {
+	pools, err := dataSource.GetResourcePools(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resource pools: %w", err)
+	}
+
+	// Loop over the set of resource pools and insert (or update) as needed
+	for _, pool := range pools {
+		dataChangeEvent, err := persistObjectWithChangeEvent(
+			ctx, c.repository.Db, pool, pool.ResourcePoolID, nil, func(object interface{}) any {
+				record, _ := object.(models.ResourcePool)
+				return models.ResourcePoolToModel(&record)
+			})
+		if err != nil {
+			return nil, fmt.Errorf("failed to persist resource pool: %w", err)
```

Comment:
I think we should try persist as many pools as we can. Instead of returning an error, we could log (similar to what you do in GetResourcePools) and continue. We can remove that pool from pools (again similar to GetResourcePools).

Reply:
I'll revisit when we add more functionality.